server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java

```diff
@@ -184,6 +220,20 @@ public GroupShardsIterator<ShardIterator> searchShards(
         return GroupShardsIterator.sortAndCreate(new ArrayList<>(set));
     }
 
+    private void setWeightedRoundRobinAttributes(ClusterState clusterState, ClusterService clusterService) {
+        WeightedRoundRobinRoutingMetadata weightedRoundRobinRoutingMetadata = clusterState.metadata()
+            .custom(WeightedRoundRobinRoutingMetadata.TYPE);
+        this.isWeightedRoundRobinEnabled = weightedRoundRobinRoutingMetadata != null;
+        if (this.isWeightedRoundRobinEnabled) {
+            this.wrrWeights = weightedRoundRobinRoutingMetadata.getWrrWeight();
+            this.wrrShardsCache = getWrrShardsCache() != null ? getWrrShardsCache() : new WRRShardsCache(clusterService);
```

Comment:
Can we please have a singleton instance of `WRRShardCache` to be injected maybe via `ClusterService` which on `ClusterService#doClose` closes the  `WRRShardCache`

Reply:
Made changes.